### PR TITLE
Removed enum for ExceptionType

### PR DESCRIPTION
**Pull request by brentschmaltz**
The enums that map to an Exception to create have been removed. This creates a dependency in the code, benchmarks show very little change.

Reverted change related to LastKnownGood in that modified the release call graph.
Introduced a model for a derived ExceptionDetail coordinated with an additional ctor for Exceptions.

### CODE REVIEW

**Review by FuPingFranco — APPROVED** [review]
LGTM

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs [hunk 1]
```diff
@@ -150,7 +150,7 @@ await ValidateJWEAsync(jsonWebToken, validationParameters, currentConfiguration,
                 return result;
             }
 
-            if (TokenUtilities.IsRecoverableErrorType(result.UnwrapError().Type))
+            if (TokenUtilities.IsRecoverableException(result.UnwrapError().GetException()))
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs:153]: Why are we allocating the exception when we could simply check the type and save the allocation?

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs [hunk 2]
```diff
@@ -182,13 +182,13 @@ await ValidateJWEAsync(jsonWebToken, validationParameters, currentConfiguration,
                 {
                     validationParameters.RefreshBeforeValidation = false;
                     validationParameters.ValidateWithLKG = true;
-                    ExceptionType recoverableExceptionType = result.UnwrapError().Type;
+                    Exception recoverableExceptionType = result.UnwrapError().GetException();
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs:185]: Same question around allocating the exception.

File: src/Microsoft.IdentityModel.Tokens/Validation/Results/Details/ExceptionDetail.cs [hunk 3]
```diff
@@ -53,19 +57,60 @@ public ExceptionDetail(
         /// Creates an instance of an <see cref="Exception"/> using <see cref="ExceptionDetail"/>
         /// </summary>
         /// <returns>An instantance of an Exception.</returns>
-        public Exception GetException()
+        public virtual Exception GetException()
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.Tokens/Validation/Results/Details/ExceptionDetail.cs:60]: The Activator used in this PR: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2800, is more extensible than this approach. Why are we resorting to still using a bunch of ifs checking for the different exceptions?

One more thing to consider is that the `ArgumentNullException` not inheriting from `SecurityTokenException` and therefore doesn't have a stack trace.

**Comment by iNinja** [comment]
This PR is re-doing the work done in a previous PR, https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2800.

I am not seeing the point on merging these changes over the linked PR. I am happy to take the approach to initialise exceptions with the ExceptionDetail presented here onto that one.

**Comment by brentschmaltz** [comment]
Added changes to https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2801.

Will pull in additional work around adding ExceptionDetails to SecurityTokenException.

**Pull request closed without merging**